server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java

```diff
@@ -0,0 +1,181 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.qualityprofile;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.MapDifference;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.sonar.api.ServerComponent;
+import org.sonar.api.rule.RuleKey;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.qualityprofile.db.QualityProfileDto;
+import org.sonar.core.util.NonNullInputFunction;
+import org.sonar.server.db.DbClient;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+public class QProfileComparator implements ServerComponent {
+
+  private final DbClient dbClient;
+
+  private final QProfileLoader profileLoader;
+
+  public QProfileComparator(DbClient dbClient, QProfileLoader profileLoader) {
+    this.dbClient = dbClient;
+    this.profileLoader = profileLoader;
+  }
+
+  public QProfileComparisonResult compare(String leftKey, String rightKey) {
+    QProfileComparisonResult result = new QProfileComparisonResult();
+
+    DbSession dbSession = dbClient.openSession(false);
+
+    try {
+      compare(leftKey, rightKey, dbSession, result);
+    } finally {
+      dbSession.close();
+    }
+
+    return result;
+  }
+
+  private void compare(String leftKey, String rightKey, DbSession session, QProfileComparisonResult result) {
+    result.left = dbClient.qualityProfileDao().getByKey(session, leftKey);
+    Preconditions.checkArgument(result.left != null, String.format("Could not find left profile '%s'", leftKey));
+    result.right = dbClient.qualityProfileDao().getByKey(session, rightKey);
+    Preconditions.checkArgument(result.right != null, String.format("Could not find right profile '%s'", leftKey));
+
+    Map<RuleKey, ActiveRule> leftActiveRulesByRuleKey = loadActiveRules(leftKey);
+    Map<RuleKey, ActiveRule> rightActiveRulesByRuleKey = loadActiveRules(rightKey);
+
+    Set<RuleKey> allRules = Sets.newHashSet();
+    allRules.addAll(leftActiveRulesByRuleKey.keySet());
+    allRules.addAll(rightActiveRulesByRuleKey.keySet());
+
+    for (RuleKey ruleKey : allRules) {
+      if (!leftActiveRulesByRuleKey.containsKey(ruleKey)) {
+        result.inRight.put(ruleKey, rightActiveRulesByRuleKey.get(ruleKey));
+      } else if (!rightActiveRulesByRuleKey.containsKey(ruleKey)) {
+        result.inLeft.put(ruleKey, leftActiveRulesByRuleKey.get(ruleKey));
+      } else {
+        compare(leftActiveRulesByRuleKey.get(ruleKey), rightActiveRulesByRuleKey.get(ruleKey), result);
+      }
+    }
+  }
+
+  private void compare(ActiveRule leftRule, ActiveRule rightRule, QProfileComparisonResult result) {
+    RuleKey key = leftRule.key().ruleKey();
+    if (leftRule.params().equals(rightRule.params()) && leftRule.severity().equals(rightRule.severity())) {
```

Comment:
Are you sure comparison of hashmap with equals() is working when order of param is different ?